File: sdks/python/engram_client/resources/events.py
"""Realtime and SSE streaming resource mixin."""

from __future__ import annotations

import json
from typing import Any, AsyncIterator

import httpx

from ..errors import EngramError
from .base import ResourceMixin


class EventsMixin(ResourceMixin):
    """Realtime Server-Sent Events (SSE) streaming and progress notification operations."""

    def parse_sse_event(self, event_chunk: str) -> dict[str, Any] | None:
        """Parse a raw SSE event payload into a dictionary."""
        lines = event_chunk.strip().split("\n")
        event_type = "memory_created"
        data_str = ""
        seq_id = None

        for line in lines:
            if line.startswith("event:"):
                event_type = line[6:].strip()
            elif line.startswith("id:"):
                raw_id = line[3:].strip()
                try:
                    seq_id = int(raw_id)
                except ValueError:
                    pass
            elif line.startswith("data:"):
                data_str = line[5:].strip()

        if not data_str:
            return None

        try:
            parsed = json.loads(data_str)
            if not isinstance(parsed, dict):
                return None
            return {
                "seq_id": seq_id or parsed.get("seq_id"),
                "type": event_type,
                "timestamp": parsed.get("timestamp"),
                "memory_id": parsed.get("memory_id"),
                "preview": parsed.get("preview"),
                "changes": parsed.get("changes"),
                "data": parsed.get("data", parsed),
            }
        except Exception:
            return None

    async def stream_events(
        self,
        *,
        workspace: str | None = None,
        event_types: str | list[str] | None = None,
        last_event_id: int | str | None = None,
    ) -> AsyncIterator[dict[str, Any]]:
        """Stream Server-Sent Events from `GET /v1/events` as an async generator.

        Args:
            workspace: Optional workspace filter.
            event_types: Event type or list of event types to subscribe to.
            last_event_id: Optional last sequence ID for resumable stream replay.
        """
        client = getattr(self, "_client", None)
        if client is None:
            raise EngramError("EngramClient is closed")

        params: dict[str, str] = {}
        if workspace is not None:
            params["workspace"] = workspace
        if event_types is not None:
            if isinstance(event_types, list):
                params["event_types"] = ",".join(event_types)
            else:
                params["event_types"] = str(event_types)

        headers: dict[str, str] = {"Accept": "text/event-stream"}
        if last_event_id is not None:
            headers["Last-Event-Id"] = str(last_event_id)

        try:
            async with client.stream("GET", "/v1/events", params=params, headers=headers) as response:
                if response.status_code != 200:
                    raise EngramError(f"SSE stream failed with status {response.status_code}")

                buffer = ""
                async for line in response.aiter_lines():
                    if line == "":
                        if buffer:
                            event = self.parse_sse_event(buffer)
                            if event:
                                yield event
                            buffer = ""
                    else:
                        if not line.startswith(":"):
                            buffer += line + "\n"
        except httpx.HTTPError as err:
            raise EngramError(f"Streaming connection failed: {err}") from err

    async def watch_progress(
        self,
        token: str | int,
        *,
        workspace: str | None = None,
    ) -> AsyncIterator[dict[str, Any]]:
        """Stream progress notifications matching a specific progressToken.

        Args:
            token: The progressToken integer or string.
            workspace: Optional workspace filter.
        """
        str_token = str(token)
        async for event in self.stream_events(workspace=workspace, event_types="progress"):
            if event.get("type") == "progress":
                data = event.get("data", {})
                event_token = str(data.get("progress_token", data.get("progressToken", "")))
                if event_token == str_token:
                    yield {
                        "seq_id": event.get("seq_id"),
                        "type": "progress",
                        "timestamp": event.get("timestamp"),
                        "preview": event.get("preview"),
                        "data": {
                            "progress_token": event_token,
                            "progress": data.get("progress", 0),
                            "total": data.get("total"),
                            "message": data.get("message", event.get("preview")),
                            "workspace": data.get("workspace"),
                        },
                    }

File: sdks/typescript/src/resources/events.ts
import { BaseResource } from "./base.js";
import type {
  ProgressEvent,
  RealtimeEvent,
  RealtimeEventType,
  StreamEventsOptions,
} from "../types.js";

export class EventsResource extends BaseResource {
  /**
   * Parse a raw Server-Sent Events line/chunk into a structured RealtimeEvent.
   */
  parseEvent(eventChunk: string): RealtimeEvent | null {
    const lines = eventChunk.split("\n");
    let eventType: RealtimeEventType = "memory_created";
    let dataStr = "";
    let id: number | undefined;

    for (const line of lines) {
      if (line.startsWith("event:")) {
        eventType = line.slice(6).trim() as RealtimeEventType;
      } else if (line.startsWith("id:")) {
        const parsedId = Number.parseInt(line.slice(3).trim(), 10);
        if (!Number.isNaN(parsedId)) {
          id = parsedId;
        }
      } else if (line.startsWith("data:")) {
        dataStr = line.slice(5).trim();
      }
    }

    if (!dataStr) {
      return null;
    }

    try {
      const parsed = JSON.parse(dataStr) as Record<string, unknown>;
      return {
        seqId: id ?? (parsed.seq_id as number | undefined),
        type: eventType,
        timestamp: (parsed.timestamp as string) ?? new Date().toISOString(),
        memoryId: parsed.memory_id as number | undefined,
        preview: parsed.preview as string | undefined,
        changes: parsed.changes as string[] | undefined,
        data: (parsed.data as Record<string, unknown>) ?? parsed,
      };
    } catch {
      return null;
    }
  }

  /**
   * Consume Server-Sent Events from `GET /v1/events` as an AsyncIterable.
   */
  async *stream(
    baseUrl: string,
    headers: Record<string, string>,
    options?: StreamEventsOptions
  ): AsyncIterable<RealtimeEvent> {
    const params = new URLSearchParams();
    if (options?.eventTypes) {
      const types = Array.isArray(options.eventTypes)
        ? options.eventTypes.join(",")
        : options.eventTypes;
      params.set("event_types", types);
    }
    if (options?.workspace) {
      params.set("workspace", options.workspace);
    }

    const query = params.toString();
    const url = `${baseUrl}/v1/events${query ? `?${query}` : ""}`;

    const reqHeaders: Record<string, string> = {
      ...headers,
      Accept: "text/event-stream",
    };
    if (options?.lastEventId !== undefined) {
      reqHeaders["Last-Event-Id"] = String(options.lastEventId);
    }

    const response = await fetch(url, {
      method: "GET",
      headers: reqHeaders,
      signal: options?.signal,
    });

    if (!response.ok) {
      throw new Error(`SSE stream failed with status ${response.status}`);
    }

    if (!response.body) {
      return;
    }

    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";

    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });
        const parts = buffer.split("\n\n");
        buffer = parts.pop() ?? "";

        for (const part of parts) {
          const trimmed = part.trim();
          if (!trimmed || trimmed.startsWith(":")) {
            continue; // Skip comments/keepalives
          }
          const event = this.parseEvent(trimmed);
          if (event) {
            yield event;
          }
        }
      }
    } finally {
      reader.releaseLock();
    }
  }

  /**
   * Watch for progress notifications matching a specific progressToken.
   */
  async watchProgress(
    baseUrl: string,
    headers: Record<string, string>,
    token: string | number,
    onProgress: (event: ProgressEvent) => void,
    signal?: AbortSignal
  ): Promise<() => void> {
    const controller = new AbortController();
    const combinedSignal = signal ?? controller.signal;

    (async () => {
      try {
        const eventStream = this.stream(baseUrl, headers, {
          eventTypes: ["progress"],
          signal: combinedSignal,
        });

        for await (const event of eventStream) {
          if (event.type === "progress" && event.data) {
            const dataToken =
              event.data.progress_token ?? event.data.progressToken;
            if (String(dataToken) === String(token)) {
              onProgress({
                seqId: event.seqId,
                type: "progress",
                timestamp: event.timestamp,
                preview: event.preview,
                data: {
                  progressToken: dataToken as string | number,
                  progress: (event.data.progress as number) ?? 0,
                  total: event.data.total as number | undefined,
                  message: (event.data.message as string) ?? event.preview,
                  workspace: event.data.workspace as string | undefined,
                },
              });
            }
          }
        }
      } catch {
        // Stream aborted or closed
      }
    })();

    return () => controller.abort();
  }
}
